README
sts-empire has been created from vms-empire. for the first release,
this file has not been changed aside from this text

		README for vms-empire

VMS-Empire is a simulation of a full-scale war between two
emperors, the computer and you.  Naturally, there is only
room for one, so the object of the game is to destroy
the other.  The computer plays by the same rules that you
do.  This game is the ancestor of all the multiplayer 4X
simulations out there, including Civilization and Master of Orion.

This code originated under VMS but has been forward-ported to
modern ANSI C.  It is known to run not only on Linux but
also under Windows and Mac OS X; all you need is a local
port of the curses library.

The files HACKING and BUGS will be of interest if you intend
to modify this code.

The ancestral code was by Chuck Simmons <chuck@amdahl.com>.  Since
around 1990 it has been maintained by Eric S. Raymond <esr@thyrsus.com>


NEWS
1.00 2021-12-11 
       vms-empire forked into sts-empire and released as version 1.00



			vms-empire news

1.15: 2021-02-25 20:52:1Z
	Minor documentation fices and code cleanup.

1.15: 2020-01-26 01:20:45Z
	Fix for infinite-fighter-tange bug,
	New 'J' command to show city production in editing mode.
        Rehosted on GitLab.

1.14: 2016-01-05 06:49:17Z
        Add an identifying magic cookie and version to save files.
        Full ANSI prototypes for private functions as well as public.
	Markup fix for probability table.

1.13: 2014-12-14T17:06:26Z
	Comment on entry to and exit from edit mode.
	Cleaned up packaging and install/uninstall productions.

1.12: 2014-05-20T09:03:09Z
	Incorporated Dennis Pinckard's fix for a Mac OS/X port bug.
	Added desktop file.

1.11: 2013-12-31T17:31:04Z
	Clean up code to modern C, apply cppcheck.
	Fix code typo reported as Debian bug #593434

1.10: 2013-08-22T23:47:21Z
	Fix core dump when a satellite bounced off the world edge.
	Full ANSIfication of function prototypes (which fixed the 
	sat bug - the code appears to have been tickling some obscure
	bug in the compilation of function calls with K&R-style prototypes).

1.9: 2012-02-16T22:59:22Z
	Drop RPM packaging. Documentation cleanups.

1.8: 2010-10-21T12:47:33Z
	Joe Wells's patch to enable different-sized worlds.  Minor fixes to
	map functions by Andrew Morrow. More on the history of the game.
	Enable arrow keys in edit mode.
	License changed from Chuck Simmons's cosmetically tweaked clone
	of the GPL to GPL proper, so the classifier systems that forge
	and archive sites use will be less confused.

1.7: 2004-09-27T21:53:41Z
	Documentation fixes.  The old empire.doc file has been replaced
	by the man page.

1.6: 2003-12-29T11:11:53Z
	Source RPMs can be built by non-root user.

1.5: 2003-12-29T11:06:21Z
	Cleanup release with new build machinery.

1.4: 2002-08-01T20:04:09Z
	Packaging tweaks.  RPM spec fixes.  READ.ME broken up into
	READ.ME and NEWS.  Obsolete bug notes removed.  Option -f to
	set savefile added.  I'd like to move the documentation to XML, 
	but can't yet because db2man doesn't render tables.

1.3: 2002-04-19T10:25:43Z
	Walt Stoneburner's cleanup patch addressing all compiler errors and 
	warnings, both in debug and in production mode; works with GCC v3.0.3.

1.2: 2000-07-28T16:04:26Z
	The victory-odds table in previous versions was seriously buggy.
	I folded in corrections from Michael Self.  I also took changes
	from James T. Jordan <kermyt@earthling.net>, who wrote some
	speedups, added ANSI prototypes, and cleaned up code.

1.1: 1995-12-06T14:02:31
	Packaging tweaks.

1.0: 1994-12-01T17:15:43Z
	I colorized and speed-tuned this and added a
	save-interval option. 

The rest of this history is Chuck Simmons's original notes.  Because
there are so many other Empire games out there now (all descended from
this one!), I've reverted to the VMS Empire name.

					-- Eric S. Raymond

History

	Apparently, this game was originally written outside of Digital,
	probably at a university.  The game was ported to DEC's VAX/VMS
	from the TOPS-10/20 FORTRAN sources available around fall 1979.
	Ed James got hold of the sources at Berkeley and converted
	portions of the code to C, mostly to use curses for the screen
	handling.  He published his modified sources on the net in
	December 1986.  Because this game ran on VMS machines for so
        long, a previous version is known as VMS Empire.

	In early 1987 I reverse engineered the program and wrote a
	version completely written in C.  In doing this, I used lots
	of structures and defined constants, and I attempted to make
	the code flexible and easy to modify.  The algorithms used
	in this C version are completely new, the names of the commands
	have been changed to be more mnemonic, and new commands have
	been implemented.  Only the format of the display is the same.
        I suspect that many of my changes are slower and less
        intelligently implemented than the originals.  Also, I have
	not implemented some of the original functionality.
	However, my hope is that the commented C sources I have written
	will prove far easier to modify and enhance than the original
        FORTRAN sources.  If you make changes for the better, by
        all means send Ed James and I a copy.

	The basic game has been heavily modified.  I've changed the
	types of objects built, modified the parameters on others,
	and added lots of new kinds of movement functions.  Read
	the man page for a complete description.

	The file 'bugs' contains lots of ideas for enhancements,
	and describes the bugs I haven't been able to find.

Final Notes

	Unfortunately, I have a rather powerful mainframe at my
	disposal which is somewhere between 10 and 40 times as
	fast as a 68020 based computer.  This means I can afford
	to use extremely inefficient algorithms.  I suspect that
	running this program on a smaller machine, such as a Sun
	workstation or Vax will not be overly rewarding.  In particular,
	the computer will take a very long time to move its pieces,
	and it may not be desirable to save the game after every move.
	(You mean your system doesn't write out 1/2 megabyte files in a
	few milliseconds?)  This second problem is easily fixed, but
	I don't yet have any good ideas for fixing the first problem.

	The size of a saved file can be easily tuned by reducing the
	LIST_SIZE constant in empire.h.  The only current simple tweak
	for making the computer move faster is to reduce the size
	of a map.

Chuck Simmons
amdahl!chuck

Ed James
edjames@ic.berkeley.edu
ucbvax!edjames

	The low-end PCs of 2002 are more powerful than Chuck's 1987 mainframe.
	Performance tuning isn't the issue it once was. :-)

	My changes enable color on machines with terminfo color support, for
	a dramatic improvement in appearance and readability of the display.
	Color support, if present, will be auto-detected at compilation time.


main.c
/*
 *    Copyright (C) 1987, 1988 Chuck Simmons
 * 
 * See the file COPYING, distributed with empire, for restriction
 * and warranty information.
 */

/*
main.c -- parse command line for empire

options:

    -w water: percentage of map that is water.  Must be in the range
              10..90.  Default is 70.
	      
    -s smooth: amount of smoothing performed to generate map.  Must
	       be a nonnegative integer.  Default is 5.
	       
    -d delay:  number of milliseconds to delay between output.
               default is 500 (0.5 seconds).

    -S saveinterval: sets turn interval between saves.
	       default is 10
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include "empire.h"
#include "extern.h"

#define OPTFLAGS "w:s:d:S:f:"

int main(argc, argv)
int argc;
char *argv[];
{
    int c;
    extern char *optarg;
    extern int optind;
    int errflg = 0;
    int wflg, sflg, dflg, Sflg;
    int land;
	
    wflg = 70; /* set defaults */
    sflg = 5;
    dflg = 500;
    Sflg = 10;
    game.savefile = "empsave.dat";

    /*
     * extract command line options
     */

    while ((c = getopt (argc, argv, OPTFLAGS)) != EOF) {
	switch (c) {
	case 'w':
	    wflg = atoi (optarg);
	    break;
	case 's':
	    sflg = atoi (optarg);
	    break;
	case 'd':
	    dflg = atoi (optarg);
	    break;
	case 'S':
	    Sflg = atoi (optarg);
	    break;
	case 'f':
	    game.savefile = optarg;
	    break;
	case '?': /* illegal option? */
	    errflg++;
	    break;
	}
    }
    if (errflg || (argc-optind) != 0) {
	(void) printf ("empire: usage: empire [-w water] [-s smooth] [-d delay]\n");
	exit (1);
    }

    if (wflg < 10 || wflg > 90) {
	(void) printf ("empire: -w argument must be in the range 10..90.\n");
	exit (1);
    }
    if (sflg < 0) {
	(void) printf ("empire: -s argument must be greater or equal to zero.\n");
	exit (1);
    }
	
    if (dflg < 0 || dflg > 30000) {
	(void) printf ("empire: -d argument must be in the range 0..30000.\n");
	exit (1);
    }

    game.SMOOTH = sflg;
    game.WATER_RATIO = wflg;
    game.delay_time = dflg;
    game.save_interval = Sflg;

    /* compute min distance between cities */
    land = MAP_SIZE * (100 - game.WATER_RATIO) / 100; /* available land */
    land /= NUM_CITY; /* land per city */
    game.MIN_CITY_DIST = isqrt (land); /* distance between cities */

    empire (); /* call main routine */
    return (0);
}


empire.c
/*
 *    Copyright (C) 1987, 1988 Chuck Simmons
 * 
 * See the file COPYING, distributed with empire, for restriction
 * and warranty information.
 */

/*
empire.c -- this file contains initialization code, the main command
parser, and the simple commands.
*/

#include <stdio.h>
#include "empire.h"
#include "extern.h"

gamestate_t game;

void c_examine(void), c_movie(void);

/*
 * 03a 01Apr88 aml .Hacked movement algorithms for computer.
 * 02b 01Jun87 aml .First round of bug fixes.
 * 02a 01Jan87 aml .Translated to C.
 * 01b 27May85 cal .Fixed round number update bug. Made truename simple.
 * 01a 01Sep83 cal .Taken from a Decus tape
 */

void
empire(void)
{
    void do_command(char);
    void print_zoom(view_map_t *);

    char order;
    int turn = 0;

    ttinit (); /* init tty */
    rndini (); /* init random number generator */

    clear_screen (); /* nothing on screen */
    pos_str (7, 0, "STS-EMPIRE, Version 1.00  12-Dec-2021 Based off VMS-EMPIRE");
    pos_str (8, 0, "Wargame simulation. Detailed directions are in sts-empire manual page");
    pos_str (9, 0, "This program comes with ABSOLUTELY NO WARRANTY");
    pos_str (10, 0, "This is free software, and you are welcome to redistribute it under");
    pos_str (11, 0, "certain conditions. Distributed under GPLv3. See COPYING for more details\n");
    (void) redisplay ();

    if (!restore_game ()) /* try to restore previous game */
	init_game (); /* otherwise init a new game */

    /* Command loop starts here. */

    for (;;) { /* until user quits */
	if (game.automove) { /* don't ask for cmd in auto mode */
	    user_move ();
	    comp_move (1);
	    if (++turn % game.save_interval == 0)
		save_game ();
	}
	else {
	    prompt (""); /* blank top line */
	    redisplay();
	    prompt ("Your orders? ");
	    order = get_chx (); /* get a command */
	    do_command (order);
	}
    }
}

/*
Execute a command.
*/

void
do_command(char orders)
{
    void c_debug(char order), c_quit(void), c_sector(void), c_map(void);
    void c_give(void);

    char e;
    int ncycle;

    switch (orders) {
    case 'A': /* turn on auto move mode */
	game.automove = true;
	error ("Now in Auto-Mode");
	user_move ();
	comp_move (1);
	save_game ();
	break;

    case 'C': /* give a city to the computer */
	c_give ();
	break;
	
    case 'D': /* display round number */
	error ("Round #%d", game.date);
	break;

    case 'E': /* examine enemy map */
	if (game.resigned) c_examine ();
	else huh (); /* illegal command */
	break;

    case 'F': /* print map to file */
	c_map ();
	break;

    case 'G': /* give one free enemy move */
	comp_move (1);
	break;

    case 'H': /* help */
	help (help_cmd, cmd_lines);
	break;

    case 'J': /* edit mode */
	ncycle = cur_sector ();
	if (ncycle == -1) ncycle = 0;
	edit (sector_loc (ncycle));
	break;

    case 'M': /* move */
	user_move ();
	comp_move (1);
	save_game ();
	break;

    case 'N': /* give enemy free moves */
	ncycle = getint ("Number of free enemy moves: ");
	comp_move (ncycle);
	save_game ();
	break;

    case 'P': /* print a sector */
	c_sector ();
	break;

    case '\026': /* some interrupt */
    case 'Q': /* quit */
	c_quit ();
	break;

    case 'R': /* restore game */
	clear_screen ();
	e = restore_game ();
	break;

    case 'S': /* save game */
	save_game ();
	break;
	
    case 'T': /* trace: toggle game.save_movie flag */
	game.save_movie = !game.save_movie;
	if (game.save_movie) comment ("Saving movie screens to 'empmovie.dat'.");
	else comment ("No longer saving movie screens.");
	break;

    case 'W': /* watch movie */
	if (game.resigned || game.debug) replay_movie ();
	else error ("You cannot watch movie until computer resigns.");
	break;

    case 'Z': /* print compressed map */
	print_zoom (game.user_map);
	break;

    case '\014': /* redraw the screen */
	redraw ();
	break;

    case '+': /* change debug state */
	e = get_chx();
	if ( e  ==  '+' )
	    game.debug = true;
	else if ( e  ==  '-' )
	    game.debug = false;
	else huh ();
	break;

    default:
	if (game.debug)
	    c_debug (orders); /* debug */
	else
	    huh (); /* illegal command */
	break;
    }
}

/*
Give an unowned city (if any) to the computer.  We make
a list of unowned cities, choose one at random, and mark
it as the computers.
*/

void
c_give(void)
{
    int unowned[NUM_CITY];
    count_t i, count;

    count = 0; /* nothing in list yet */
    for (i = 0; i < NUM_CITY; i++) {
	if (game.city[i].owner == UNOWNED) {
	    unowned[count] = i; /* remember this city */
	    count += 1;
	}
    }
    if (count == 0) {
	error ("There are no unowned cities.");
	ksend ("There are no unowned cities.");
	return;
    }
    i = irand (count);
    i = unowned[i]; /* get city index */
    game.city[i].owner = COMP;
    game.city[i].prod = NOPIECE;
    game.city[i].work = 0;
    scan (game.comp_map, game.city[i].loc);
}

/*
Debugging commands should be implemented here.  
The order cannot be any legal command.
*/

void
c_debug(char order)
{
    char e;

    switch (order) {
    case '#' : c_examine (); break;
    case '%' : c_movie (); break;
	
    case '@': /* change trace state */
	e = get_chx();
	if ( e  ==  '+' )
	    game.trace_pmap = true;
	else if ( e  ==  '-' )
	    game.trace_pmap = false;
	else
	    huh ();
	break;

    case '$': /* change game.print_debug state */
	e = get_chx();
	if ( e  ==  '+' )
	    game.print_debug = true;
	else if ( e  ==  '-' )
	    game.print_debug = false;
	else
	    huh ();
	break;

    case '&': /* change game.print_vmap state */
	game.print_vmap = get_chx();
	break;

    default: huh (); break;
    }
}

/*
The quit command.  Make sure the user really wants to quit.
*/

void
c_quit(void)
{
    if (getyn ("QUIT - Are you sure? ")) {
	empend ();
    }
}

/*
Print a sector.  Read the sector number from the user
and print it.
*/

void
c_sector(void)
{
    int num;

    num = get_range ("Sector number? ", 0, NUM_SECTORS-1);
    print_sector_u (num);
}

/*
Print the map to a file.  We ask for a filename, attempt to open the
file, and if successful, print out the user's information to the file.
We print the map sideways to make it easier for the user to print
out the map.
*/

void
c_map(void)
{
    FILE *f;
    int i, j;
    char line[MAP_HEIGHT+2];

    prompt ("Filename? ");
    get_str (game.jnkbuf, STRSIZE);

    f = fopen (game.jnkbuf, "w");
    if (f == NULL) {
	error ("I can't open that file.");
	return;
    }
    for (i = 0; i < MAP_WIDTH; i++) { /* for each column */
	for (j = MAP_HEIGHT-1; j >= 0; j--) { /* for each row */
	    line[MAP_HEIGHT-1-j] = game.user_map[row_col_loc(j,i)].contents;
	}
	j = MAP_HEIGHT-1;
	while (j >= 0 && line[j] == ' ') /* scan off trailing blanks */
	    j -= 1;
			
	line[++j] = '\n';
	line[++j] = 0; /* trailing null */
	(void) fputs (line, f);
    }
    (void) fclose (f);
}

/*
Allow user to examine the computer's map.
*/

void
c_examine(void)
{
    int num;

    num = get_range ("Sector number? ", 0, NUM_SECTORS-1);
    print_sector_c (num);
}

/*
We give the computer lots of free moves and
Print a "zoomed" version of the computer's map.
*/

void
c_movie(void)
{
    for (;;) {
	comp_move (1);
	print_zoom (game.comp_map);
	save_game ();
#ifdef PROFILE
	if (game.date == 125) empend();
#endif
    }
}

/* end */


data.c
/*
 *    Copyright (C) 1987, 1988 Chuck Simmons
 * 
 * See the file COPYING, distributed with empire, for restriction
 * and warranty information.
 */

/*
Static data.

One of our hopes is that changing the types of pieces that
exist in the game is mostly a matter of modifying this file.
However, see also the help routine, empire.h, and empire.6.
*/

#include "empire.h"

/*
Piece attributes.  Notice that the Transport is allowed only one hit.
In the previous version of this game, the user could easily win simply
by building armies and troop transports.  We attempt to alleviate this
problem by making transports far more fragile.  We have also increased
the range of a fighter from 20 to 30 so that fighters will be somewhat
more useful.
*/

piece_attr_t piece_attr[] = {
	{'A', /* character for printing piece */
	 "army", /* name of piece */ 
	 "army", /* nickname */
	 "an army", /* name with preceding article */
	 "armies", /* plural */
	 "+", /* terrain */
	 5, /* units to build */
	 1, /* strength */
	 1, /* max hits */
	 1, /* movement */
	 0, /* capacity */
	 INFINITY}, /* range */

	/*
	 For fighters, the range is set to an even multiple of the speed.
	 This allows user to move fighter, say, two turns out and two
	 turns back.
	*/
	 
	{'F', "fighter", "fighter", "a fighter", "fighters",
		".+", 10, 1,  1, 8, 0, 32},

	{'P', "patrol boat", "patrol", "a patrol boat", "patrol boats",
		".",  15, 1,  1, 4, 0, INFINITY},
		
	{'D', "destroyer", "destroyer", "a destroyer", "destroyers",
		".",  20, 1,  3, 2, 0, INFINITY},

	{'S', "submarine", "submarine", "a submarine", "submarines",
		".",  20, 3,  2, 2, 0, INFINITY},

	{'T', "troop transport", "transport", "a troop transport", "troop transports",
		".",  30, 1,  1, 2, 6, INFINITY},

	{'C', "aircraft carrier", "carrier", "an aircraft carrier", "aircraft carriers",
		".",  30, 1,  8, 2, 8, INFINITY},

	{'B', "battleship", "battleship", "a battleship", "battleships",
		".",  40, 2, 10, 2, 0, INFINITY},
		
	{'Z', "satellite", "satellite", "a satellite", "satellites",
		".+", 50, 0, 1, 10, 0, 500}
};

/* Direction offsets. */

int dir_offset [] = {-MAP_WIDTH, /* north */
		     -MAP_WIDTH+1, /* northeast */
		     1, /* east */
		     MAP_WIDTH+1, /* southeast */
		     MAP_WIDTH, /* south */
		     MAP_WIDTH-1, /* southwest */
		     -1, /* west */
		     -MAP_WIDTH-1}; /* northwest */

/* Names of movement functions. */

char *func_name[] = {"none", "random", "sentry", "fill", "land",
			"explore", "load", "attack", "load", "repair",
			"transport",
			"W", "E", "D", "C", "X", "Z", "A", "Q"};

/* The order in which pieces should be moved. */
int move_order[] = {SATELLITE, TRANSPORT, CARRIER, BATTLESHIP, 
		    PATROL, SUBMARINE, DESTROYER, ARMY, FIGHTER};

/* types of pieces, in declared order */
char type_chars[] = "AFPDSTCBZ";

/* Lists of attackable objects if object is adjacent to moving piece. */

char tt_attack[] = "T";
char army_attack[] = "O*TACFBSDP";
char fighter_attack[] = "TCFBSDPA";
char ship_attack[] = "TCBSDP";

/* Define various types of objectives */

move_info_t tt_explore = { /* water objectives */
	COMP, /* home city */
	" ", /* objectives */
	{1} /* weights */
};
move_info_t tt_load = { /* land objectives */
	COMP, "$",           {1}
};

/*
Rationale for 'tt_unload':

     Any continent with four or more cities is extremely attractive,
and we should grab it quickly.  A continent with three cities is
fairly attractive, but we are willing to go slightly out of our
way to find a better continent.  Similarily for two cities on a
continent.  At one city on a continent, things are looking fairly
unattractive, and we are willing to go quite a bit out of our way
to find a better continent.

     Cities marked with a '0' are on continents where we already
have cities, and these cities will likely fall to our armies anyway,
so we don't need to dump armies on them unless everything else is
real far away.  We would prefer to use unloading transports for
taking cities instead of exploring, but we are willing to explore
if interesting cities are too far away.

     It has been suggested that continents containing one city
are not interesting.  Unfortunately, most of the time what the
computer sees is a single city on a continent next to lots of
unexplored territory.  So it dumps an army on the continent to
explore the continent and the army ends up attacking the city
anyway.  So, we have decided we might as well send the tt to
the city in the first place and increase the speed with which
the computer unloads its tts.
*/

move_info_t tt_unload     = {
	COMP, "9876543210 ", {1, 1, 1, 1, 1, 1, 11, 21, 41, 101, 61}
};

/*
 '$' represents loading tt must be first
 'x' represents tt producing city
 '0' represnets explorable territory
*/
move_info_t army_fight    = { /* land objectives */
	COMP, "O*TA ",       {1, 1, 1, 1, 11}
};
move_info_t army_load     = { /* water objectives */
	COMP, "$x",          {1, W_TT_BUILD}
};

move_info_t fighter_fight = {
	COMP, "TCFBSDPA ",   {1, 1, 5, 5, 5, 5, 5, 5, 9}
};
move_info_t ship_fight    = {
	COMP, "TCBSDP ",     {1, 1, 3, 3, 3, 3, 21}
};
move_info_t ship_repair   = {
	COMP, "X",           {1}
};

move_info_t user_army        = {
	USER, " ",   {1}
};
move_info_t user_army_attack = {
	USER, "*Xa ", {1, 1, 1, 12}
};
move_info_t user_fighter     = {
	USER, " ",   {1}
};
move_info_t user_ship        = {
	USER, " ",   {1}
};
move_info_t user_ship_repair = {
	USER, "O",   {1}
};

/*
Various help texts.
*/

char *help_cmd[] = {
	"COMMAND MODE",
	"Auto:     enter automove mode",
	"City:     give city to computer",
	"Date:     print round",
	"Examine:  examine enemy map",
	"File:     print map to file",
	"Give:     give move to computer",
	"Help:     display this text",
	"J:        enter edit mode",
	"Move:     make a move",
	"N:        give N moves to computer",
	"Print:    print a sector",
	"Quit:     quit game",
	"Restore:  restore game",
	"Save:     save game",
	"Trace:    save movie in empmovie.dat",
	"Watch:    watch movie",
	"Zoom:     display compressed map",
	"<ctrl-L>: redraw screen"
};
int cmd_lines = 19;

char *help_user[] = {
	"USER MODE",
	"QWE",
	"A D       movement directions",
	"ZXC",
	"<space>:  sit",
	"Build:    change city production",
	"Fill:     set func to fill",
	"GoExplore:set func to explore",
	"Help:     display this text",
	"I <dir>:  set func to dir",
	"J:        enter edit mode",
	"K:        set func to awake",
	"Land:     set func to land",
	"Out:      leave automove mode",
	"Print:    redraw screen",
	"Random:   set func to random",
	"Sentry:   set func to sentry",
	"Transport:set func to transport",
	"Upgrade:  set func to repair",
	"V <piece> <func>:  set city func",
	"Y:        set func to attack",
	"<ctrl-L>: redraw screen",
	"?:        describe piece"
};
int user_lines = 22;
	
char *help_edit[] = {
	"EDIT MODE",
	"QWE",
	"A D       movement directions",
	"ZXC",
	"Build:    change city production",
	"Fill:     set func to fill",
	"GoExplore:set func to explore",
	"Help:     display this text",
	"I <dir>:  set func to dir",
	"J:        show all city's production",
	"K:        set func to awake",
	"Land:     set func to land",
	"Mark:     mark piece",
	"N:        set dest for marked piece",
	"Out:      exit edit mode",
	"Print:    print sector",
	"Random:   set func to random",
	"Sentry:   set func to sentry",
	"Upgrade:  set func to repair",
	"V <piece> <func>:  set city func",
	"Y:        set func to attack",
	"<ctrl-L>: redraw screen",
	"?:        describe piece"
};
int edit_lines = 22;
